core/sqf/src/seatrans/hbase-trx/src/main/java/org/apache/hadoop/hbase/coprocessor/transactional/TrxRegionEndpoint.java.tmpl

```diff
@@ -7335,6 +7338,245 @@ TrxTransactionState isTransactionOnCommittedList(final long transactionId)
+              (startKey == null || startKey.length == 0) &&  // first region only
+              totalEntries == reader.getEntries()) {  // first file only
+
+            // Trafodion column qualifiers are ordinal numbers, but are represented
+            // as varying length unsigned little-endian integers in lexicographical
+            // order. So, for example, in a table with 260 columns, the column
+            // qualifiers (if present) will be read in this order: 
+            // 1 (x'01'), 257 (x'0101'), 2 (x'02'), 258 (x'0201'), 3 (x'03'),
+            // 259 (x'0301'), 4 (x'04'), 260 (x'0401'), 5 (x'05'), 6 (x'06'), 
+            // 7 (x'07'), ...
+            // We have crossed the boundary to the next row if and only if the
+            // next qualifier read is less than or equal to the previous, 
+            // compared unsigned, lexicographically.
+
+            HFileScanner scanner = reader.getScanner(false, false, false);
+            scanner.seekTo();  //position at beginning of first data block
+
+            // the next line should succeed, as we know the HFile is non-empty
+            Cell currKv = scanner.getKeyValue();
+            while ((more) && (currKv.getTypeByte() != KeyValue.Type.Put.getCode())) {
+              nonPutKVsSampled++;
+              more = scanner.next();
+              currKv = scanner.getKeyValue();
+            }
+            if (more) {
+              // now we have the first KeyValue in the HFile
+
+              int putKVsThisRow = 1;
+              putKVsSampled++;
+              sampleRowCount++;  // we have at least one row
+              more = scanner.next();
+    
+              while ((more) && (sampleRowCount <= ROWS_TO_SAMPLE)) {
+                Cell nextKv = scanner.getKeyValue();
+                if (nextKv.getTypeByte() == KeyValue.Type.Put.getCode()) {
+                  if (isQualifierLessThanOrEqual(nextKv,currKv)) {
+                    // we have crossed a row boundary
+                    sampleRowCount++;
+                    missingKVsCount += (numCols - putKVsThisRow);
+                    putKVsThisRow = 1;
+                  } else {
+                    putKVsThisRow++;
+                  }
+                  currKv = nextKv;
+                  putKVsSampled++;
+                } else {
+                  nonPutKVsSampled++;  // don't count these toward the number
+                } 
+              more = scanner.next();
+              }
+            }   
+  
+            if (sampleRowCount > ROWS_TO_SAMPLE) {
+              // we read one KeyValue beyond the ROWS_TO_SAMPLE-eth row, so
+              // adjust counts for that
+              putKVsSampled--;
+              sampleRowCount--;
+            }
+          }  // code for first file
+        } catch (IOException exp1) {
+          if (LOG.isErrorEnabled()) LOG.error("Trafodion estimate row count encountered exception ", exp1);
+          t = exp1;
+        } finally {
+          reader.close(false);
```

Comment:
HFileScanner doesn't have a close method until HBase 2.0 unfortunately. In the rework, though, I'll add a comment.